client/asset/btc/electrum_client.go

```diff
@@ -0,0 +1,1289 @@
+			medianTimeBlocks, height, hdrsRes.Count)
+	}
+	if hdrsRes.Count == 0 {
+		return time.Time{}, nil, errors.New("no headers retrieved")
+	}
+
+	hdrReader := hex.NewDecoder(strings.NewReader(hdrsRes.HexConcat))
+
+	var lastHdr *wire.BlockHeader
+	timestamps := make([]int64, 0, hdrsRes.Count)
+	for i := int64(0); i < int64(hdrsRes.Count); i++ {
+		hdr := &wire.BlockHeader{}
+		err = hdr.Deserialize(hdrReader)
+		if err != nil {
+			if i > 0 {
+				ew.log.Errorf("Failed to deserialize header for block %d: %v",
+					startHeight+i, err)
+				break // we have at least one time stamp, work with it
+			}
+			return time.Time{}, nil, err
+		}
+		timestamps = append(timestamps, hdr.Timestamp.Unix())
+		if i+startHeight == height {
+			lastHdr = hdr
+		}
+	}
+	// Naive way fetching each header separately:
+	// for i := height; i > height-medianTimeBlocks && i > 0; i-- {
+	// 	hdr, err := ew.getBlockHeaderByHeight(height)
+	// 	if err != nil {
+	// 		return time.Time{}, err
+	// 	}
+	// 	timestamps = append(timestamps, hdr.Timestamp.Unix())
+	// }
+
+	sort.Slice(timestamps, func(i, j int) bool {
+		return timestamps[i] < timestamps[j]
+	})
+
+	medianTimestamp := timestamps[len(timestamps)/2]
+	return time.Unix(medianTimestamp, 0), lastHdr, nil
+}
+
+func (ew *electrumWallet) getBlockHash(height int64) (*chainhash.Hash, error) {
+	hdr, err := ew.getBlockHeaderByHeight(height)
+	if err != nil {
+		return nil, err
+	}
+	hash := hdr.BlockHash()
+	return &hash, nil
+}
+
+func (ew *electrumWallet) getBestBlockHash() (*chainhash.Hash, error) {
+	inf, err := ew.wallet.GetInfo()
+	if err != nil {
+		return nil, err
+	}
+	return ew.getBlockHash(inf.SyncHeight)
+}
+
+func (ew *electrumWallet) getBestBlockHeight() (int32, error) {
+	inf, err := ew.wallet.GetInfo()
+	if err != nil {
+		return 0, err
+	}
+	return int32(inf.SyncHeight), nil
+}
+
+func (ew *electrumWallet) getBestBlockHeader(falseMedianTime ...bool) (*blockHeader, error) {
+	inf, err := ew.wallet.GetInfo()
+	if err != nil {
+		return nil, err
+	}
+	falseMedian := len(falseMedianTime) > 0 && falseMedianTime[0]
+
+	var hdr *wire.BlockHeader
+	var medianTime time.Time
+	if falseMedian {
+		hdr, err = ew.getBlockHeaderByHeight(inf.SyncHeight)
+		if err != nil {
+			return nil, err
+		}
+		medianTime = hdr.Timestamp
+	} else {
+		medianTime, hdr, err = ew.calcMedianTime(inf.SyncHeight)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	header := &blockHeader{
+		Hash:              hdr.BlockHash().String(),
+		Height:            inf.SyncHeight,
+		Confirmations:     1, // it's the head
+		Time:              hdr.Timestamp.Unix(),
+		MedianTime:        medianTime.Unix(),
+		PreviousBlockHash: hdr.PrevBlock.String(),
+	}
+	return header, nil
+}
+
+func (ew *electrumWallet) balances() (*GetBalancesResult, error) {
+	eBal, err := ew.wallet.GetBalance()
+	if err != nil {
+		return nil, err
+	}
+	return &GetBalancesResult{
+		Mine: Balances{
+			Trusted: eBal.Confirmed + eBal.Unconfirmed,
```

Comment:
Unconfirmed does not necessarily imply untrusted.  Since we don't have a good indication of what part of the unconfirmed balance is trusted or not, I put it here so that it will be `Available` in `(*baseWallet).Balance`

For DCR and BTC at least, we allow use of unconfirmed utxos (but preferring confirmed), so to stick with that I just did this.  The important rule for the `btc.Wallet` backends is that the result of `balances` must agree with `listUnspent` in terms of the trusted/untrusted balance and the "safe" field of `ListUnspentResult`.

So, the options are: (a) declare all unconfirmed utxos **unsafe** in both methods, diverging from how we operate with the other BTC wallet backends, (b) make no attempt at determining "safeness" (see DOGE approach that leaves SafePtr unset and uses legacyBalance with no notion of trustedness and simply combining confirmed+unconfirmed balances), or (c) make like `spvWallet` and see if we can make an [`ownsInputs` method](https://github.com/decred/dcrdex/commit/55c890c4815c8baa28476c820225d10a45e99010) without making electrumx server requests.

I think I may be able to do (c) without expensive operations, but I'm not hopeful.